From 001ba5f9bfaea285c8ec2404922350596e804c89 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Sun, 4 Aug 2024 03:59:16 +0200
Subject: [PATCH] Broadcast more packets to sink players


diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 62eeddf6e..762f9d24b 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -1102,7 +1102,16 @@ public abstract class PlayerList {
                 }
             }
         }
-
+        // KigPaper start
+        for (WorldServer world : server.worlds) {
+            if (world.dimension != i) {
+                continue;
+            }
+            for (EntityPlayer player : world.sinkPlayers.values()) {
+                player.playerConnection.sendPacket(packet);
+            }
+        }
+        // KigPaper end
     }
 
     // KigPaper start
@@ -1126,7 +1135,14 @@ public abstract class PlayerList {
                 }
             }
         }
-
+        for (WorldServer world : server.worlds) {
+            if (world.dimension != i) {
+                continue;
+            }
+            for (EntityPlayer player : world.sinkPlayers.values()) {
+                player.playerConnection.sendPacket(packet);
+            }
+        }
     }
     // KigPaper end
 
diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 3a8bb2a6d..3f2ea8d73 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -1084,6 +1084,12 @@ public class WorldServer extends World implements IAsyncTaskHandler {
             }
         }
 
+        // KigPaper start
+        for (EntityPlayer player : sinkPlayers.values()) {
+            player.playerConnection.sendPacket(new PacketPlayOutExplosion(d0, d1, d2, f, explosion.getBlocks(), null));
+        }
+        // KigPaper end
+
         return explosion;
     }
 
@@ -1219,7 +1225,11 @@ public class WorldServer extends World implements IAsyncTaskHandler {
                 entityplayer.playerConnection.sendPacket(packetplayoutworldparticles);
             }
         }
-
+        // KigPaper start
+        for (EntityPlayer player : sinkPlayers.values()) {
+            player.playerConnection.sendPacket(packetplayoutworldparticles);
+        }
+        // KigPaper end
     }
 
     public Entity getEntity(UUID uuid) {
-- 
2.45.2

