From 837c6d8493d4135a8c0e66ac040ad5ffc0caea80 Mon Sep 17 00:00:00 2001
From: Vanolex <vanolex@vanolex.net>
Date: Thu, 22 Jan 2026 19:40:53 +0100
Subject: [PATCH] Better-ProjectileHitEvent-API


diff --git a/src/main/java/net/minecraft/server/EntityArrow.java b/src/main/java/net/minecraft/server/EntityArrow.java
index 78bb8bfbe..e3c47de61 100644
--- a/src/main/java/net/minecraft/server/EntityArrow.java
+++ b/src/main/java/net/minecraft/server/EntityArrow.java
@@ -1,9 +1,10 @@
 package net.minecraft.server;
 
 import dev.rocco.kig.paper.api.event.PlayerCheckCriticalEvent;
-import dev.rocco.kig.paper.api.event.ProjectileCollideEvent;
+import dev.rocco.kig.paper.api.event.ProjectilePreHitEvent;
 import dev.rocco.kig.paper.api.particle.ProjectileEffect;
 import org.bukkit.Location;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.entity.LivingEntity;
 import org.bukkit.entity.Player;
 import org.bukkit.event.entity.EntityCombustByEntityEvent;
@@ -261,8 +262,15 @@ public class EntityArrow extends Entity implements IProjectile {
             // PaperSpigot end
 
             // KigPaper start
-            if (movingobjectposition != null && movingobjectposition.entity != null) {
-                ProjectileCollideEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition.entity);
+            IBlockData hitBlockData = null;
+            if (movingobjectposition != null) {
+                ProjectilePreHitEvent event;
+                if (movingobjectposition.entity != null) {
+                    event = CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition.entity);
+                } else {
+                    hitBlockData = this.world.getType(movingobjectposition.a());
+                    event = CraftEventFactory.callProjectilePreHitBlockEvent(this, movingobjectposition);
+                }
                 if (event.isCancelled()) movingobjectposition = null;
             }
             // KigPaper end
@@ -357,7 +365,7 @@ public class EntityArrow extends Entity implements IProjectile {
                     this.d = blockposition1.getX();
                     this.e = blockposition1.getY();
                     this.f = blockposition1.getZ();
-                    IBlockData iblockdata1 = this.world.getType(blockposition1);
+                    IBlockData iblockdata1 = hitBlockData == null ? this.world.getType(blockposition1) : hitBlockData; // KigPaper - use block before potential change by ProjectileHitBlockEvent
 
                     this.g = iblockdata1.getBlock();
                     this.h = this.g.toLegacyData(iblockdata1);
diff --git a/src/main/java/net/minecraft/server/EntityFireball.java b/src/main/java/net/minecraft/server/EntityFireball.java
index 7a6a0843d..4418c9784 100644
--- a/src/main/java/net/minecraft/server/EntityFireball.java
+++ b/src/main/java/net/minecraft/server/EntityFireball.java
@@ -2,7 +2,7 @@ package net.minecraft.server;
 
 import java.util.List;
 
-import dev.rocco.kig.paper.api.event.ProjectileCollideEvent;
+import dev.rocco.kig.paper.api.event.ProjectilePreHitEvent;
 import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
 
 public abstract class EntityFireball extends Entity {
@@ -128,8 +128,13 @@ public abstract class EntityFireball extends Entity {
             }
 
             // KigPaper start
-            if (movingobjectposition != null && movingobjectposition.entity != null) {
-                ProjectileCollideEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition.entity);
+            if (movingobjectposition != null) {
+                ProjectilePreHitEvent event;
+                if (movingobjectposition.entity != null) {
+                    event = CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition.entity);
+                } else {
+                    event = CraftEventFactory.callProjectilePreHitBlockEvent(this, movingobjectposition);
+                }
                 if (event.isCancelled()) movingobjectposition = null;
             }
             // KigPaper end
diff --git a/src/main/java/net/minecraft/server/EntityFishingHook.java b/src/main/java/net/minecraft/server/EntityFishingHook.java
index e5d2f80f6..d9dee525f 100644
--- a/src/main/java/net/minecraft/server/EntityFishingHook.java
+++ b/src/main/java/net/minecraft/server/EntityFishingHook.java
@@ -1,8 +1,9 @@
 package net.minecraft.server;
 
-import dev.rocco.kig.paper.api.event.ProjectileCollideEvent;
+import dev.rocco.kig.paper.api.event.ProjectilePreHitEvent;
 import dev.rocco.kig.paper.api.particle.ProjectileEffect;
 import org.bukkit.Location;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.entity.Fish;
 import org.bukkit.entity.Player;
 import org.bukkit.event.player.PlayerFishEvent;
@@ -216,8 +217,15 @@ public class EntityFishingHook extends Entity {
             // PaperSpigot end
 
             // KigPaper start
-            if (movingobjectposition != null && movingobjectposition.entity != null) {
-                ProjectileCollideEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition.entity);
+            boolean isBlockEventCancelled = false;
+            if (movingobjectposition != null) {
+                ProjectilePreHitEvent event;
+                if (movingobjectposition.entity != null) {
+                    event = CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition.entity);
+                } else {
+                    event = CraftEventFactory.callProjectilePreHitBlockEvent(this, movingobjectposition);
+                    isBlockEventCancelled = event.isCancelled();
+                }
                 if (event.isCancelled()) movingobjectposition = null;
             }
             // KigPaper end
@@ -240,7 +248,16 @@ public class EntityFishingHook extends Entity {
                         trailEffect.spawn(loc);
                     }
                 }
+
+
+                boolean oldNoclip = this.noclip; // KigPaper
+                this.noclip = this.noclip || isBlockEventCancelled; // KigPaper - allow projectiles to pass through blocks when event cancelled
+
                 this.move(this.motX, this.motY, this.motZ);
+
+                this.noclip = oldNoclip; // KigPaper
+
+
                 float f1 = MathHelper.sqrt(this.motX * this.motX + this.motZ * this.motZ);
 
                 this.yaw = (float) (MathHelper.b(this.motX, this.motZ) * 180.0D / 3.1415927410125732D);
diff --git a/src/main/java/net/minecraft/server/EntityProjectile.java b/src/main/java/net/minecraft/server/EntityProjectile.java
index 793cd1191..7692de18b 100644
--- a/src/main/java/net/minecraft/server/EntityProjectile.java
+++ b/src/main/java/net/minecraft/server/EntityProjectile.java
@@ -2,9 +2,10 @@ package net.minecraft.server;
 
 // KigPaper start
 
-import dev.rocco.kig.paper.api.event.ProjectileCollideEvent;
+import dev.rocco.kig.paper.api.event.ProjectilePreHitEvent;
 import dev.rocco.kig.paper.api.particle.ProjectileEffect;
 import org.bukkit.Location;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.entity.Player;
 
 import java.util.List;
@@ -179,8 +180,11 @@ public abstract class EntityProjectile extends Entity implements IProjectile {
         // PaperSpigot end
 
         // KigPaper start
-        if (movingobjectposition != null && movingobjectposition.entity != null) {
-            ProjectileCollideEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition.entity);
+        if (movingobjectposition != null) {
+            ProjectilePreHitEvent event =
+                    movingobjectposition.entity != null ?
+                    CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition.entity) :
+                    CraftEventFactory.callProjectilePreHitBlockEvent(this, movingobjectposition);
             if (event.isCancelled()) movingobjectposition = null;
         }
         // KigPaper end
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index ff62ce025..5b495c4ea 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -10,6 +10,8 @@ import com.google.common.base.Function;
 import com.google.common.base.Functions;
 
 import dev.rocco.kig.paper.api.event.ProjectileCollideEvent;
+import dev.rocco.kig.paper.api.event.ProjectilePreHitBlockEvent;
+import dev.rocco.kig.paper.api.event.ProjectilePreHitEntityEvent;
 import net.minecraft.server.*;
 
 import org.bukkit.Bukkit;
@@ -766,11 +768,29 @@ public class CraftEventFactory {
     }
 
     // KigPaper start
+
+    /**
+     * @deprecated use {@link ProjectilePreHitEntityEvent}
+     */
+    @Deprecated
     public static ProjectileCollideEvent callProjectileCollideEvent(Entity entity, Entity collided) {
         ProjectileCollideEvent event = new ProjectileCollideEvent((Projectile) entity.getBukkitEntity(), collided.getBukkitEntity());
         entity.world.getServer().getPluginManager().callEvent(event);
         return event;
     }
+
+    public static ProjectilePreHitEntityEvent callProjectilePreHitEntityEvent(Entity projectile, Entity hitEntity) {
+        ProjectilePreHitEntityEvent event = new ProjectilePreHitEntityEvent((Projectile) projectile, hitEntity.getBukkitEntity());
+        projectile.world.getServer().getPluginManager().callEvent(event);
+        return event;
+    }
+
+    public static ProjectilePreHitBlockEvent callProjectilePreHitBlockEvent(Entity projectile, MovingObjectPosition movingObjectPosition) {
+        org.bukkit.block.Block block = projectile.world.getWorld().getBlockAt(movingObjectPosition.a().getX(), movingObjectPosition.a().getY(), movingObjectPosition.a().getZ());
+        ProjectilePreHitBlockEvent event = new ProjectilePreHitBlockEvent((Projectile) projectile.getBukkitEntity(), block, CraftBlock.notchToBlockFace(movingObjectPosition.direction));
+        projectile.world.getServer().getPluginManager().callEvent(event);
+        return event;
+    }
     // KigPaper end
 
     public static ExpBottleEvent callExpBottleEvent(Entity entity, int exp) {
-- 
2.51.0.windows.1

