From 4c047c71642425ee68ef92da18ca04692a0a3417 Mon Sep 17 00:00:00 2001
From: Vanolex <vanolex@vanolex.net>
Date: Wed, 21 Jan 2026 01:33:51 +0100
Subject: [PATCH] Better-ProjectileHitEvent-API


diff --git a/src/main/java/net/minecraft/server/EntityArrow.java b/src/main/java/net/minecraft/server/EntityArrow.java
index 78bb8bfbe..eab1b7550 100644
--- a/src/main/java/net/minecraft/server/EntityArrow.java
+++ b/src/main/java/net/minecraft/server/EntityArrow.java
@@ -1,9 +1,10 @@
 package net.minecraft.server;
 
 import dev.rocco.kig.paper.api.event.PlayerCheckCriticalEvent;
-import dev.rocco.kig.paper.api.event.ProjectileCollideEvent;
+import dev.rocco.kig.paper.api.event.ProjectilePreHitEvent;
 import dev.rocco.kig.paper.api.particle.ProjectileEffect;
 import org.bukkit.Location;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.entity.LivingEntity;
 import org.bukkit.entity.Player;
 import org.bukkit.event.entity.EntityCombustByEntityEvent;
@@ -153,6 +154,9 @@ public class EntityArrow extends Entity implements IProjectile {
 
     public void t_() {
         super.t_();
+
+        IBlockData hitBlockData = null; // KigPaper
+
         if (this.lastPitch == 0.0F && this.lastYaw == 0.0F) {
             float f = MathHelper.sqrt(this.motX * this.motX + this.motZ * this.motZ);
 
@@ -232,9 +236,22 @@ public class EntityArrow extends Entity implements IProjectile {
                 }
             }
 
+            ProjectilePreHitEvent event1; // KigPaper
+
             if (entity != null) {
-                movingobjectposition = new MovingObjectPosition(entity);
+                // KigPaper start
+                event1 = CraftEventFactory.callProjectileCollideEvent(this, entity);
+                if (!event1.isCancelled()) movingobjectposition = new MovingObjectPosition(entity);
+                // KigPaper end
+            }
+
+            // KigPaper start
+            if (movingobjectposition != null && movingobjectposition.entity == null) {
+                hitBlockData = this.world.getType(movingobjectposition.a());
+                event1 = CraftEventFactory.callProjectilePreHitBlockEvent(this, movingobjectposition);
+                if (event1.isCancelled()) movingobjectposition = null;
             }
+            // KigPaper end
 
             if (movingobjectposition != null && movingobjectposition.entity != null && movingobjectposition.entity instanceof EntityHuman) {
                 EntityHuman entityhuman = (EntityHuman) movingobjectposition.entity;
@@ -260,12 +277,7 @@ public class EntityArrow extends Entity implements IProjectile {
             // KigPaper end
             // PaperSpigot end
 
-            // KigPaper start
-            if (movingobjectposition != null && movingobjectposition.entity != null) {
-                ProjectileCollideEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition.entity);
-                if (event.isCancelled()) movingobjectposition = null;
-            }
-            // KigPaper end
+            // KigPaper - moved ProjectileCollideEvent
 
             if (movingobjectposition != null) {
                 org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileHitEvent(this); // CraftBukkit - Call event
@@ -357,7 +369,7 @@ public class EntityArrow extends Entity implements IProjectile {
                     this.d = blockposition1.getX();
                     this.e = blockposition1.getY();
                     this.f = blockposition1.getZ();
-                    IBlockData iblockdata1 = this.world.getType(blockposition1);
+                    IBlockData iblockdata1 = hitBlockData == null ? this.world.getType(blockposition1) : hitBlockData; // KigPaper - use block before potential change by ProjectileHitBlockEvent
 
                     this.g = iblockdata1.getBlock();
                     this.h = this.g.toLegacyData(iblockdata1);
diff --git a/src/main/java/net/minecraft/server/EntityFireball.java b/src/main/java/net/minecraft/server/EntityFireball.java
index 7a6a0843d..983775a28 100644
--- a/src/main/java/net/minecraft/server/EntityFireball.java
+++ b/src/main/java/net/minecraft/server/EntityFireball.java
@@ -3,8 +3,12 @@ package net.minecraft.server;
 import java.util.List;
 
 import dev.rocco.kig.paper.api.event.ProjectileCollideEvent;
+import dev.rocco.kig.paper.api.event.ProjectilePreHitBlockEvent;
+import dev.rocco.kig.paper.api.event.ProjectilePreHitEntityEvent;
+import dev.rocco.kig.paper.api.event.ProjectilePreHitEvent;
 import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
 
+
 public abstract class EntityFireball extends Entity {
 
     private int e = -1;
@@ -123,14 +127,21 @@ public abstract class EntityFireball extends Entity {
                 }
             }
 
+            ProjectilePreHitEvent event; // KigPaper
+
             if (entity != null) {
-                movingobjectposition = new MovingObjectPosition(entity);
+                // KigPaper start
+                event = CraftEventFactory.callProjectileCollideEvent(this, entity);
+                if (!event.isCancelled()) movingobjectposition = new MovingObjectPosition(entity);
+                // KigPaper end
             }
 
             // KigPaper start
-            if (movingobjectposition != null && movingobjectposition.entity != null) {
-                ProjectileCollideEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition.entity);
-                if (event.isCancelled()) movingobjectposition = null;
+            if (movingobjectposition != null && movingobjectposition.entity == null) {
+                event = CraftEventFactory.callProjectilePreHitBlockEvent(this, movingobjectposition);
+                if (event.isCancelled()) {
+                    movingobjectposition = null;
+                }
             }
             // KigPaper end
 
diff --git a/src/main/java/net/minecraft/server/EntityFishingHook.java b/src/main/java/net/minecraft/server/EntityFishingHook.java
index e5d2f80f6..735f1e3b6 100644
--- a/src/main/java/net/minecraft/server/EntityFishingHook.java
+++ b/src/main/java/net/minecraft/server/EntityFishingHook.java
@@ -1,8 +1,11 @@
 package net.minecraft.server;
 
-import dev.rocco.kig.paper.api.event.ProjectileCollideEvent;
+import dev.rocco.kig.paper.api.event.ProjectilePreHitBlockEvent;
+import dev.rocco.kig.paper.api.event.ProjectilePreHitEntityEvent;
+import dev.rocco.kig.paper.api.event.ProjectilePreHitEvent;
 import dev.rocco.kig.paper.api.particle.ProjectileEffect;
 import org.bukkit.Location;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.entity.Fish;
 import org.bukkit.entity.Player;
 import org.bukkit.event.player.PlayerFishEvent;
@@ -198,9 +201,26 @@ public class EntityFishingHook extends Entity {
                 }
             }
 
+            ProjectilePreHitEvent event; // KigPaper
+
             if (entity != null) {
-                movingobjectposition = new MovingObjectPosition(entity);
+                // KigPaper start
+                event = CraftEventFactory.callProjectileCollideEvent(this, entity);
+                if (!event.isCancelled()) movingobjectposition = new MovingObjectPosition(entity);
+                // KigPaper end
+            }
+
+            // KigPaper start
+            // KigPaper - allow clipping though ground when ProjectileHitBlockEvent is cancelled
+            boolean isBlockEventCancelled = false;
+            if (movingobjectposition != null && movingobjectposition.entity == null) {
+                event = CraftEventFactory.callProjectilePreHitBlockEvent(this, movingobjectposition);
+                if (event.isCancelled()) {
+                    isBlockEventCancelled = true;
+                    movingobjectposition = null;
+                }
             }
+            // KigPaper end
 
             // PaperSpigot start - Allow fishing hooks to fly through vanished players the shooter can't see
             if (movingobjectposition != null && movingobjectposition.entity instanceof EntityPlayer && owner != null && owner instanceof EntityPlayer) {
@@ -215,12 +235,7 @@ public class EntityFishingHook extends Entity {
             // KigPaper end
             // PaperSpigot end
 
-            // KigPaper start
-            if (movingobjectposition != null && movingobjectposition.entity != null) {
-                ProjectileCollideEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition.entity);
-                if (event.isCancelled()) movingobjectposition = null;
-            }
-            // KigPaper end
+            // KigPaper - moved ProjectileCollideEvent call
 
             if (movingobjectposition != null) {
                 org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileHitEvent(this); // Craftbukkit - Call event
@@ -240,7 +255,16 @@ public class EntityFishingHook extends Entity {
                         trailEffect.spawn(loc);
                     }
                 }
+
+
+                boolean oldNoclip = this.noclip; // KigPaper
+                this.noclip = this.noclip || isBlockEventCancelled; // KigPaper - allow projectiles to pass through blocks when event cancelled
+
                 this.move(this.motX, this.motY, this.motZ);
+
+                this.noclip = oldNoclip; // KigPaper
+
+
                 float f1 = MathHelper.sqrt(this.motX * this.motX + this.motZ * this.motZ);
 
                 this.yaw = (float) (MathHelper.b(this.motX, this.motZ) * 180.0D / 3.1415927410125732D);
diff --git a/src/main/java/net/minecraft/server/EntityProjectile.java b/src/main/java/net/minecraft/server/EntityProjectile.java
index 793cd1191..eb9ceede3 100644
--- a/src/main/java/net/minecraft/server/EntityProjectile.java
+++ b/src/main/java/net/minecraft/server/EntityProjectile.java
@@ -2,9 +2,10 @@ package net.minecraft.server;
 
 // KigPaper start
 
-import dev.rocco.kig.paper.api.event.ProjectileCollideEvent;
+import dev.rocco.kig.paper.api.event.ProjectilePreHitEvent;
 import dev.rocco.kig.paper.api.particle.ProjectileEffect;
 import org.bukkit.Location;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.entity.Player;
 
 import java.util.List;
@@ -135,6 +136,8 @@ public abstract class EntityProjectile extends Entity implements IProjectile {
             vec3d1 = new Vec3D(movingobjectposition.pos.a, movingobjectposition.pos.b, movingobjectposition.pos.c);
         }
 
+        ProjectilePreHitEvent event; // KigPaper
+
         if (!this.world.isClientSide) {
             Entity entity = null;
             List list = this.world.getEntities(this, this.getBoundingBox().a(this.motX, this.motY, this.motZ).grow(1.0D, 1.0D, 1.0D));
@@ -161,9 +164,18 @@ public abstract class EntityProjectile extends Entity implements IProjectile {
             }
 
             if (entity != null) {
-                movingobjectposition = new MovingObjectPosition(entity);
+                // KigPaper start
+                event = CraftEventFactory.callProjectileCollideEvent(this, entity);
+                if (!event.isCancelled()) movingobjectposition = new MovingObjectPosition(entity);
+                // KigPaper end
             }
         }
+        // KigPaper start
+        if (movingobjectposition != null && movingobjectposition.entity == null) {
+            event = CraftEventFactory.callProjectilePreHitBlockEvent(this, movingobjectposition);
+            if (event.isCancelled()) movingobjectposition = null;
+        }
+        // KigPaper end
 
         // PaperSpigot start - Allow projectiles to fly through vanished players the shooter can't see
         if (movingobjectposition != null && movingobjectposition.entity instanceof EntityPlayer && shooter != null && shooter instanceof EntityPlayer) {
@@ -178,12 +190,7 @@ public abstract class EntityProjectile extends Entity implements IProjectile {
         // KigPaper end
         // PaperSpigot end
 
-        // KigPaper start
-        if (movingobjectposition != null && movingobjectposition.entity != null) {
-            ProjectileCollideEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition.entity);
-            if (event.isCancelled()) movingobjectposition = null;
-        }
-        // KigPaper end
+        // KigPaper - moved ProjectileCollideEvent call
 
         if (movingobjectposition != null) {
             if (movingobjectposition.type == MovingObjectPosition.EnumMovingObjectType.BLOCK && this.world.getType(movingobjectposition.a()).getBlock() == Blocks.PORTAL) {
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index ff62ce025..5b495c4ea 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -10,6 +10,8 @@ import com.google.common.base.Function;
 import com.google.common.base.Functions;
 
 import dev.rocco.kig.paper.api.event.ProjectileCollideEvent;
+import dev.rocco.kig.paper.api.event.ProjectilePreHitBlockEvent;
+import dev.rocco.kig.paper.api.event.ProjectilePreHitEntityEvent;
 import net.minecraft.server.*;
 
 import org.bukkit.Bukkit;
@@ -766,11 +768,29 @@ public class CraftEventFactory {
     }
 
     // KigPaper start
+
+    /**
+     * @deprecated use {@link ProjectilePreHitEntityEvent}
+     */
+    @Deprecated
     public static ProjectileCollideEvent callProjectileCollideEvent(Entity entity, Entity collided) {
         ProjectileCollideEvent event = new ProjectileCollideEvent((Projectile) entity.getBukkitEntity(), collided.getBukkitEntity());
         entity.world.getServer().getPluginManager().callEvent(event);
         return event;
     }
+
+    public static ProjectilePreHitEntityEvent callProjectilePreHitEntityEvent(Entity projectile, Entity hitEntity) {
+        ProjectilePreHitEntityEvent event = new ProjectilePreHitEntityEvent((Projectile) projectile, hitEntity.getBukkitEntity());
+        projectile.world.getServer().getPluginManager().callEvent(event);
+        return event;
+    }
+
+    public static ProjectilePreHitBlockEvent callProjectilePreHitBlockEvent(Entity projectile, MovingObjectPosition movingObjectPosition) {
+        org.bukkit.block.Block block = projectile.world.getWorld().getBlockAt(movingObjectPosition.a().getX(), movingObjectPosition.a().getY(), movingObjectPosition.a().getZ());
+        ProjectilePreHitBlockEvent event = new ProjectilePreHitBlockEvent((Projectile) projectile.getBukkitEntity(), block, CraftBlock.notchToBlockFace(movingObjectPosition.direction));
+        projectile.world.getServer().getPluginManager().callEvent(event);
+        return event;
+    }
     // KigPaper end
 
     public static ExpBottleEvent callExpBottleEvent(Entity entity, int exp) {
-- 
2.51.0.windows.1

